text/unify-theorem-definition-fixpoint-syntax.md

```diff
@@ -0,0 +1,193 @@
+  annotation (in the form of an attribute).  This also opens the way
+  to supporting (non-mutually) recursive `#[ recursive ] Theorem`, and
+  `#[ recursive ] Definition` as an equivalent to `Fixpoint` to define
+  (possibly mutually) recursive objects.
+
+# Detailed design
+
+The current syntax is:
+
+- for `Definition` / `Example`:
+  ```
+  {| Definition | Example } @ident_decl @def_body
+
+  def_body ::= {* @binder } {? : @type } := {? @reduce } @term
+             | {* @binder } : @type
+  ```
+
+- for `Fixpoint`:
+  ```
+  Fixpoint @fix_def {* with @fix_def }
+
+  fix_def ::= @ident_decl {* @binder } {? @fixannot } {? : @type } {? := @term } {? @decl_notations }
+  ```
+
+- for `CoFixpoint`:
+  ```
+  CoFixpoint @cofix_def {* with @cofix_def }
+
+  cofix_def ::= @ident_decl {* @binder } {? : @type } {? := @term } {? @decl_notations }
+  ```
+
+- for `Theorem` and variants:
+  ```
+  @thm_token @ident_decl {* @binder } : @type {* with @ident_decl {* @binder } : @type }
+
+  thm_token ::= Theorem
+              | Lemma
+              | Fact
+              | Remark
+              | Corollary
+              | Proposition
+              | Property
+  ```
+
+The new syntax would be:
+
+```
+@decl_token @decl_def {* with @decl_def }
+
+decl_token := Definition
+            | Example
+            | Fixpoint
+            | CoFixpoint
+            | Theorem
+            | Lemma
+            | Fact
+            | Remark
+            | Corollary
+            | Proposition
+            | Property
+
+decl_def := @ident_decl {* @binder } {? @fixannot } @decl_body {? @decl_notations }
+
+decl_body := {? : @type } := {? @reduce } @term
+           | : @type
+```
+
+with the following newly supported attributes:
+
+- `recursive` and `corecursive` to declare a fixpoint / cofixpoint:
+
+  - attribute `recursive` is implicit for `Fixpoint` and rejected by
+    `CoFixpoint`;
+  - attribute `corecursive` is implicit for `CoFixpoint` and rejected
+    by `Fixpoint`;
+  - attribute `recursive` or `corecursive` is compulsory to use the
+    `with` variant of `Definition`;
+  - using the `with` variant of `Theorem` without attribute
+    `recursive` or `corecursive` raises a deprecation message and is
+    interpreted as if `recursive` had been used;
+  - the `@fixannot` clause requires the `recursive` attribute
+    being used (or implicit in the case of `Fixpoint`).
+
+- `opaque` and `transparent` to declare an object opaque or
+  transparent like `Qed` and `Defined` do (note that this is not the
+  same as what the `Opaque` and `Transparent` commands do):
+
+  - `transparent` is assumed by default for `Definition`, `Fixpoint`,
+    `CoFixpoint` but `opaque` is accepted;
+  - `opaque` is assumed by default for `Theorem` and variants but
+    `transparent` is accepted.
+  - these attributes are only accepted when a body is provided (when
+    no body is provided the transparency is determined by the ending
+    `Defined` or `Qed`).
```

Comment:
A fun alternative, which I don't think has been considered in the discussion yet, is to force `Defined` or `Qed` to align with the declaration attribute. It does break some code, though. But it could be considered. Then, to have a transparent theorem, you would need to do:

```coq
#[transparent] Theorem .
Proof.
…
Defined.
```

This may be useful for asynchronous evaluation. Or maybe not, I dont know.

Reply:
>  to force `Defined` or `Qed` to align with the declaration attribute

It was proposed to force Theorem to end in Qed and Definition with Defined, have the former default to opaque, the latter to transparent, and have the attributes to overrule the default. @Zimmi48 seem to like it and has a plan to slowly introduce it. It was discussed at the Coq call, sorry you could not attend.

Reply:
> sorry you could not attend

I don't usually :slightly_smiling_face: I just lurk in random CEPs.

Reply:
@aspiwack: I'll push an update today that will summarize the result of the discussion in the thread and in the call.